src/Messages/Attachments/TSAttachment.m

```diff
@@ -33,18 +34,34 @@ - (instancetype)initWithServerId:(UInt64)serverId
     return self;
 }
 
+- (BOOL)isDecimalNumberText:(NSString *)text {
+    return [text componentsSeparatedByCharactersInSet:[NSCharacterSet decimalDigitCharacterSet]].count == 1;
+}
+
 - (nullable instancetype)initWithCoder:(NSCoder *)coder
 {
     self = [super initWithCoder:coder];
     if (!self) {
         return self;
     }
 
-    if (_attachmentSchemaVersion < 2) {
```

Comment:
This is the heart of the possible fix.  

* We were doing a schema upgrade in `TSAttachment`, which is the base class of `TSAttachmentPointer` and `TSAttachmentStream`.  
* For any instance of `TSAttachmentStream`, this upgrade was inappropriate and possibly lead to data loss. Instances of `TSAttachmentStream` don't set their attachmentSchemaVersion, so they were always be "upgraded".  
* `TSAttachmentStream` represents two separate things: an outgoing attachment or an already-downloaded incoming attachment.  I don't see how any harm could be done by the latter case, but the former case is dangerous.
* An outgoing attachment's server id is set after the upload is complete.  If the `TSAttachmentStream` was reloaded before the corresponding message is sent to the service, that message will end up with an incorrect service id.
* The data loss occurs because the schema upgrade overwrites the service id by parsing the uniqueId of the attachment.  For any new attachment, the uniqueId will be a UUID.  The uniqueId is parsed using `[NSString integerValue]`.  Given a UUID value like 11BFAA01-ADD4-4E48-81B2-2654941C9ECD, this method will simply parse the prefix characters that can be parsed as an integer, in this case the prefix 11.  This would explain why the 404s so frequently have small server ids.
* `TSOutgoingMessage` doesn't store instances of `TSAttachmentStream` - it stores the uniqueIds for the message's attachments.  This would explain how the `TSAttachmentStream` would get reloaded between attachment upload completion and message upload.
* So it all fits pretty neatly.  The only problem is that even with this diagnosis I can't find a repro.  Nonetheless the current behavior was clearly wrong so we should merge this change.